frontend/lib/src/components/shared/StreamlitMarkdown/StreamlitMarkdown.tsx

```diff
@@ -193,17 +225,36 @@ export const HeadingWithAnchor: FunctionComponent<
     return React.createElement(tag, tagProps, children)
   }
 
-  return React.createElement(
+  const actionElements = (
+    <HeaderActionElements
+      elementId={elementId}
+      help={help}
+      hideAnchor={hideAnchor}
+    />
+  )
+
+  // We nest the action-elements (tooltip, link-icon) into the header element (e.g. h1),
+  // so that it appears inline. For context: we also tried setting the h's display attribute to 'inline', but
+  // then we would need to add padding to the outer container and fiddle with the vertical alignment.
+  const headerElementWithActions = React.createElement(
     tag,
-    { ...tagProps, ref, id: elementId },
-    <StyledLinkIconContainer data-testid="StyledLinkIconContainer">
-      {elementId && !hideAnchor && (
-        <StyledLinkIcon href={`#${elementId}`}>
-          <LinkIcon size="18" />
-        </StyledLinkIcon>
-      )}
-      <StyledHeaderContent>{children}</StyledHeaderContent>
-    </StyledLinkIconContainer>
+    {
+      ...tagProps,
+      ref,
+      id: elementId,
+    },
+    <>
+      {children}
+      {actionElements}
+    </>
+  )
+  return (
+    <StyledHeadingWithActionElements
+      className="stHeadingWithActionElements"
+      data-testid="stStyledHeadingWithActionElements"
```

Comment:
nit: i think it would be better if className and data-testid overlap in most cases. But do we even need to set the classname here?

Reply:
good point, I have replaced the class name here